braindecode/eegneuralnet.py

```diff
@@ -0,0 +1,153 @@
+import numpy as np
+import torch
+from sklearn.metrics import get_scorer
+from skorch import NeuralNet
+from skorch.callbacks import BatchScoring, EpochScoring, EpochTimer, PrintLog
+from skorch.utils import noop, to_numpy, train_loss_score, valid_loss_score
+
+from .training.scoring import (CroppedTimeSeriesEpochScoring,
+                               CroppedTrialEpochScoring, PostEpochTrainScoring)
+
+
+class _EEGNeuralNet:
+    def _yield_callbacks(self):
+        # Here we parse the callbacks supplied as strings,
+        # e.g. 'accuracy', to the callbacks skorch expects
+        for name, cb, named_by_user in super()._yield_callbacks():
+            if name == 'str':
+                train_cb, valid_cb = self._parse_str_callback(cb)
+                yield train_cb
+                if self.train_split is not None:
+                    yield valid_cb
+            else:
+                yield name, cb, named_by_user
+
+    def _parse_str_callback(self, cb_supplied_name):
+        scoring = get_scorer(cb_supplied_name)
+        scoring_name = scoring._score_func.__name__
+        assert scoring_name.endswith(
+            ('_score', '_error', '_deviance', '_loss'))
+        if (scoring_name.endswith('_score') or
+                cb_supplied_name.startswith('neg_')):
+            lower_is_better = False
+        else:
+            lower_is_better = True
+        train_name = f'train_{cb_supplied_name}'
+        valid_name = f'valid_{cb_supplied_name}'
+        if self.cropped:
+            # TODO: use CroppedTimeSeriesEpochScoring when time series target
+            # In case of cropped decoding we are using braindecode
+            # specific scoring created for cropped decoding
+            train_scoring = CroppedTrialEpochScoring(
+                cb_supplied_name, lower_is_better, on_train=True, name=train_name
+            )
+            valid_scoring = CroppedTrialEpochScoring(
+                cb_supplied_name, lower_is_better, on_train=False, name=valid_name
+            )
+        else:
+            train_scoring = PostEpochTrainScoring(
+                cb_supplied_name, lower_is_better, name=train_name
+            )
+            valid_scoring = EpochScoring(
+                cb_supplied_name, lower_is_better, on_train=False, name=valid_name
+            )
+        named_by_user = True
+        train_valid_callbacks = [
+            (train_name, train_scoring, named_by_user),
+            (valid_name, valid_scoring, named_by_user)
+        ]
+        return train_valid_callbacks
+
+    # pylint: disable=arguments-differ
+    def get_loss(self, y_pred, y_true, *args, **kwargs):
```

Comment:
Why do we have this method? Could we remove it?

Reply:
It was introduced to counter the skorch implementation of `NeuralNetClassifier.get_loss`. In this method https://github.com/skorch-dev/skorch/blob/35cbaa73892c42f1d5131cdd406246b618f0aecd/skorch/classifier.py#L144-L150 they apply log to `y_pred` when using `NLLoss` which in our case would break the model because we would have LogSoftmax in models and then an additional Log applied in skorch.

This is definitely not needed in the case of `EEGREgressor`.

Reply:
Removed in EEGRegressor